### PR TITLE
fix: set & update store open close time in a specific format & store open close settings make extendable for pro.

### PR DESCRIPTION
**Pull request by MdAsifHossainNadim**
### Update store open close settings

- Use filters and actions for make extendable our store open close feature in "templates/settings/store-form.php".
- Optimize dokan_store_time code and make 12 hour format data for save in "includes/Dashboard/Templates/Settings.php".
- Compare store opening/closing data in 12 hour format and update dokan_is_store_open function for extendable.

Related with dokan-pro: <a href='https://github.com/weDevsOfficial/dokan-pro/pull/1559'>https://github.com/weDevsOfficial/dokan-pro/pull/1559/</a>
fix: https://github.com/weDevsOfficial/dokan/issues/1410

### CODE REVIEW

File: includes/Dashboard/Templates/Settings.php [hunk 1]
```diff
@@ -502,19 +502,54 @@ function insert_settings_info() {
...
+                $opening_time = \DateTimeImmutable::createFromFormat( wc_time_format(), $opening_time, new \DateTimeZone( dokan_wp_timezone_string() ) );
+                $opening_time = $opening_time->format( 'g:i a' );
+
+                // Make 12 hours format data for save.
+                $closing_time = \DateTimeImmutable::createFromFormat( wc_time_format(), $closing_time, new \DateTimeZone( dokan_wp_timezone_string() ) );
+                $closing_time = $closing_time->format( 'g:i a' );
+
+                // Get and set current day's data for update dokan store time.
+                $current_day['status']        = $store_status;
+                $current_day['opening_time']  = isset( $opening_time ) ? $opening_time : '';
```
Comment (ManiruzzamanAkash) [inline includes/Dashboard/Templates/Settings.php:546]: I think there would be no possibility for `$opening_time` become unsetted in here. So, `isset()` is not necessary. Please recheck again. Same for `$closing_time`

File: includes/functions.php [hunk 2]
```diff
@@ -3219,42 +3219,109 @@ function dokan_remove_action_button( $actions ) {
...
+ *
+ * @since 3.2.16
+ *
+ * @param string $current_day
+ * @param int    $index
+ *
+ * @return mixed|string
+ */
+function dokan_get_store_opening_time( $current_day, $index = 0 ) {
+    $profile_info       = dokan_get_store_info( dokan_get_current_user_id() );
```
Comment (ManiruzzamanAkash) [inline includes/functions.php:3286]: make this variable name `$store_info`

File: includes/functions.php [hunk 3]
```diff
@@ -3219,42 +3219,109 @@ function dokan_remove_action_button( $actions ) {
...
+ * Get store closing time.
+ *
+ * @since 3.2.16
+ *
+ * @param string $current_day
+ * @param int    $index
+ *
+ * @return mixed|string
+ */
+function dokan_get_store_closing_time( $current_day, $index = 0 ) {
```
Comment (ManiruzzamanAkash) [inline includes/functions.php:3311]: These two functions implement almost same type of things. We should follow the `DRY concept`. Better would be, create just one function. And we could also generate two function from these in future.
```php

/*
* @params string $time_type, eg: `opening_time` or `closing_time`
*/
function dokan_get_store_time( $time_type, $current_day ) { // You don't use the index anywhere also, so why's that?
     
  // just some snippet would be like this - 
  $store_time = isset( $store_info['dokan_store_time'] ) ? $store_info['dokan_store_time'] : '';
  $time       = isset( $store_time[ $current_day ][$time_type] ) ? $store_time[ $current_day ][$time_type] : '';
 // do rest as this
}
```

Then in future use case - 
```php
dokan_get_store_time( 'opening_time', 1 );
dokan_get_store_time( 'closing_time', 1 );
```

File: templates/settings/store-form.php [hunk 4]
```diff
@@ -248,31 +255,66 @@
...
+                            </label>
+                            <label for="opening-time" class="time" style="visibility: <?php echo isset( $status ) && $status == 'open' ? 'visible' : 'hidden'; ?>" >
+                                <?php
+                                    /* translators: %1$s is replaced with current day string, %2$s is replaced with placeholder string & %3$s is replaced with store opening time as a string value. */
+                                    $dokan_store_opening_entities = sprintf( '<input type="text" class="dokan-form-control opening-time" name="opening_time[%1$s]" id="opening-time[%1$s]" placeholder="%2$s" value="%3$s" >', esc_attr( $day_key ), esc_attr( '00:00' ), esc_attr( dokan_get_store_opening_time( $day_key ) ) );
+
+                                    echo apply_filters( 'dokan_store_opening_time_entities', $dokan_store_opening_entities, $day_key );
+                                ?>
+                            </label>
+                            <label for="closing-time" class="time" style="visibility: <?php echo isset( $status ) && $status == 'open' ? 'visible' : 'hidden'; ?>" >
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:293]: This for will not work in `<label for="opening-time" >`

Remember, why we use this `for`. We need the id in the input same name as this. Then, on click this label, the input will be focused. 

Talk with me, if you have any confusion.

File: templates/settings/store-form.php [hunk 5]
```diff
@@ -248,31 +255,66 @@
...
+
+                                <?php
+                                /**
+                                 * @since 3.2.16
+                                 */
+                                do_action( 'after_dokan_status_selection', $status );
+                                ?>
+
+                            </label>
+                            <label for="opening-time" class="time" style="visibility: <?php echo isset( $status ) && $status == 'open' ? 'visible' : 'hidden'; ?>" >
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:285]: For `for=""` in `label` please check the below note.

File: templates/settings/store-form.php [hunk 6]
```diff
@@ -248,31 +255,66 @@
...
+
+                                    echo apply_filters( 'dokan_store_opening_time_entities', $dokan_store_opening_entities, $day_key );
+                                ?>
+                            </label>
+                            <label for="closing-time" class="time" style="visibility: <?php echo isset( $status ) && $status == 'open' ? 'visible' : 'hidden'; ?>" >
+                                <?php
+                                    /* translators: %1$s is replaced with current day string, %2$s is replaced with placeholder string & %3$s is replaced with store closing time as a string value. */
+                                    $dokan_store_closing_entities = sprintf( '<input type="text" class="dokan-form-control closing-time" name="closing_time[%1$s]" id="closing-time[%1$s]" placeholder="%2$s" value="%3$s" >', esc_attr( $day_key ), esc_attr( '00:00' ), esc_attr( dokan_get_store_closing_time( $day_key ) ) );
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:296]: Why `esc_attr` in static string - `esc_attr( '00:00' ) `?

File: templates/settings/store-form.php [hunk 7]
```diff
@@ -338,25 +380,74 @@
 
         $('#dokan-store-time-enable').trigger('change');
 
+        $( '.dokan-form-group-container' ).each( function(e) {
+            $(this).find( '.dokan-form-group' ).each( function (e) {
+                let selectedWorkingStatus = $(this).find( '.dokan-on-off option[selected]' ).length;
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:385]: make this `selectedWorkingStatus ` variable type as `const`.

File: templates/settings/store-form.php [hunk 8]
```diff
@@ -338,25 +380,74 @@
...
-            step: <?php echo 'h' === strtolower( get_option( 'time_format' ) ) ? '60' : '30'; ?>
+            timeFormat: '<?php echo addcslashes( esc_attr( wc_time_format() ), '\\' ); ?>',
+            step: <?php echo 'h:i' === strtolower( wc_time_format() ) ? '60' : '30'; ?>,
         });
+
+        // Update closing time on focus when previously setted up closing time field.
+        $( '.dokan-form-group-container' ).on( 'focus', '.dokan-form-control',  function(e) {
+            e.stopPropagation();
+
+            let self       = $(this),
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:420]: use, like `$( this )`
and
make all of the variable with `const` as, you wouldn't change these later.

File: templates/settings/store-form.php [hunk 9]
```diff
@@ -338,25 +380,74 @@
...
+                    maxTime: '<?php echo 'h:i' === strtolower( wc_time_format() ) ? '23:00' : '11:30 pm' ?>',
+                    timeFormat: '<?php echo addcslashes( esc_attr( wc_time_format() ), '\\' ); ?>',
+                    step: '<?php echo 'h:i' === strtolower( wc_time_format() ) ? '60' : '30'; ?>',
+                } );
+            }
+        } );
+
+        // Update timepicker time index after change store opening time.
+        $( '.dokan-form-group' ).on( 'change', '.opening-time', function() {
+            let self      = $(this),
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:438]: please do the same here, as top.

File: includes/Dashboard/Templates/Settings.php [hunk 10]
```diff
@@ -502,19 +502,54 @@ function insert_settings_info() {
...
-                );
+                ];
             }
 
-            $find_address      = ! empty( $post_data['find_address'] ) ? sanitize_text_field( $post_data['find_address'] ) : $default_locations['address'];
-            $default_location  = $default_locations['latitude'] . ',' . $default_locations['longitude'];
-            $location          = ! empty( $post_data['find_address'] ) ? sanitize_text_field( $post_data['location'] ) : $default_location;
+            $find_address     = ! empty( $_POST['find_address'] ) ? wc_clean( wp_unslash( $_POST['find_address'] ) ) : $default_locations['address'];
+            $default_location = $default_locations['latitude'] . ',' . $default_locations['longitude'];
+            $location         = ! empty( $_POST['find_address'] ) ? wc_clean( wp_unslash( $_POST['location'] ) ) : $default_location;
```
Comment (tanmayjay) [inline includes/Dashboard/Templates/Settings.php]: Why are you checking `empty` on `$_POST['find_address']` whereas you are using the value of `$_POST['location']`?
And also, as others above said, `wc_clean` is not needed everywhere.

File: includes/functions.php [hunk 11]
```diff
@@ -3219,42 +3219,109 @@ function dokan_remove_action_button( $actions ) {
...
+ * @param  string days
  *
  * @since  2.8.2
  *
- * @return string
+ * @return string|array
  */
-function dokan_get_translated_days( $day ) {
-    switch ( $day ) {
+function dokan_get_translated_days( $days = true ) {
```
Comment (tanmayjay) [inline includes/functions.php]: As you have stored the `key => value` pair in the variable, it's no more necessary to check all the key manually. You can rewrite the function the following way:

```
function dokan_get_translated_days( $day = '' ) {
    $all_days = array(
        'sunday'    => __( 'Sunday', 'dokan-lite' ),
        'monday'    => __( 'Monday', 'dokan-lite' ),
        'tuesday'   => __( 'Tuesday', 'dokan-lite' ),
        'wednesday' => __( 'Wednesday', 'dokan-lite' ),
        'thursday'  => __( 'Thursday', 'dokan-lite' ),
        'friday'    => __( 'Friday', 'dokan-lite' ),
        'saturday'  => __( 'Saturday', 'dokan-lite' ),
    );

    if ( empty( $day ) ) {
        return $all_days;
    }

    if ( isset( $all_days[ $day ] ) ) {
        return $all_days[ $day ];
    }

    if ( 'close' === $day ) {
        return apply_filters( 'dokan_store_close_day_label', __( 'Off Day', 'dokan-lite' ) );
    }

    return apply_filters( 'dokan_get_translated_days', '', $day );
}
```

**Review by tanmayjay — CHANGES_REQUESTED** [review]

File: includes/functions.php [hunk 12]
```diff
@@ -3219,42 +3219,84 @@ function dokan_remove_action_button( $actions ) {
...
- * @param  string day
+ * @param  string|null days
  *
  * @since  2.8.2
  *
- * @return string
+ * @return string|array
  */
-function dokan_get_translated_days( $day ) {
+function dokan_get_translated_days( $day = null ) {
```
Comment (tanmayjay) [inline includes/functions.php:3228]: As you have stored the `key => value` pair in the variable, it's no more necessary to check all the key manually. You can rewrite the function the following way:

```
function dokan_get_translated_days( $day = '' ) {
    $all_days = array(
        'sunday'    => __( 'Sunday', 'dokan-lite' ),
        'monday'    => __( 'Monday', 'dokan-lite' ),
        'tuesday'   => __( 'Tuesday', 'dokan-lite' ),
        'wednesday' => __( 'Wednesday', 'dokan-lite' ),
        'thursday'  => __( 'Thursday', 'dokan-lite' ),
        'friday'    => __( 'Friday', 'dokan-lite' ),
        'saturday'  => __( 'Saturday', 'dokan-lite' ),
    );

    if ( empty( $day ) ) {
        return $all_days;
    }

    if ( isset( $all_days[ $day ] ) ) {
        return $all_days[ $day ];
    }

    if ( 'close' === $day ) {
        return apply_filters( 'dokan_store_close_day_label', __( 'Off Day', 'dokan-lite' ) );
    }

    return apply_filters( 'dokan_get_translated_days', '', $day );
}
```

**Review by tanmayjay — CHANGES_REQUESTED** [review]

**Review by ManiruzzamanAkash — APPROVED** [review]

**Review by tanmayjay — APPROVED** [review]

File: includes/Dashboard/Templates/Settings.php [hunk 13]
```diff
@@ -502,19 +502,54 @@ function insert_settings_info() {
...
+            $dokan_days       = dokan_get_translated_days();
+            $dokan_store_time = [];
+
+            // Get & set 7 days opening & closing time for update dokan store time.
+            foreach ( $dokan_days as $day_key => $day ) {
+                $opening_time          = isset( $_POST['opening_time'][ $day_key ] ) ? wc_clean( wp_unslash( $_POST['opening_time'][ $day_key ] ) ) : '';
+                $closing_time          = isset( $_POST['closing_time'][ $day_key ] ) ? wc_clean( wp_unslash( $_POST['closing_time'][ $day_key ] ) ) : '';
+                $store_status          = ! empty( $_POST[ $day_key ]['working_status'] ) ? sanitize_text_field( wp_unslash( $_POST[ $day_key ]['working_status'] ) ) : 'close';
+                $current_day           = [];
+                $current_day['status'] = $store_status;
```
Comment (tanmayjay) [inline includes/Dashboard/Templates/Settings.php:524]: This line is not needed here.

File: includes/Dashboard/Templates/Settings.php [hunk 14]
```diff
@@ -502,19 +502,54 @@ function insert_settings_info() {
...
+
+                // If pass null value or our store is not open then our store will be close.
+                if ( empty( $opening_time ) || is_array( $opening_time ) || empty( $closing_time ) || is_array( $closing_time ) || 'open' !== $store_status ) {
+                    $current_day['status']        = 'close';
+                    $current_day['opening_time']  = '';
+                    $current_day['closing_time']  = '';
+                    $dokan_store_time[ $day_key ] = $current_day;
```
Comment (tanmayjay) [inline includes/Dashboard/Templates/Settings.php:531]: You could just reduce this as below:
```
                    $dokan_store_time[ $day_key ] = [
                        'status'       => 'close',
                        'opening_time' => '',
                        'closing_time' => '',
                    ];
```

File: includes/Dashboard/Templates/Settings.php [hunk 15]
```diff
@@ -502,19 +502,54 @@ function insert_settings_info() {
...
+
+                // Make 12 hours format data for save.
+                $closing_time = \DateTimeImmutable::createFromFormat( wc_time_format(), $closing_time, new \DateTimeZone( dokan_wp_timezone_string() ) );
+                $closing_time = $closing_time->format( 'g:i a' );
+
+                // Get and set current day's data for update dokan store time.
+                $current_day['status']        = $store_status;
+                $current_day['opening_time']  = $opening_time;
+                $current_day['closing_time']  = $closing_time;
+                $dokan_store_time[ $day_key ] = $current_day;    // Make dokan store time data here.
```
Comment (tanmayjay) [inline includes/Dashboard/Templates/Settings.php:548]: Mentioned above to reduce variables.

**Review by tanmayjay — CHANGES_REQUESTED** [review]

File: includes/functions.php [hunk 16]
```diff
@@ -3219,42 +3219,77 @@ function dokan_remove_action_button( $actions ) {
...
+ * @param int    $index
+ *
+ * @return mixed|string
+ */
+function dokan_get_store_times( $current_day, $times_type, $index = 0 ) {
+    $store_info         = dokan_get_store_info( dokan_get_current_user_id() );
+    $dokan_store_time   = isset( $store_info['dokan_store_time'] ) ? $store_info['dokan_store_time'] : '';
+    $dokan_closing_time = isset( $dokan_store_time[ $current_day ][ $times_type ] ) ? $dokan_store_time[ $current_day ][ $times_type ] : '';
+
+    if ( ! empty( $dokan_closing_time ) && is_array( $dokan_closing_time ) ) {
```
Comment (tanmayjay) [inline includes/functions.php:3284]: Rewrite this part as below:
```
        if ( empty( $dokan_closing_time ) ) {
            return '';
        }

        if ( ! is_array( $dokan_closing_time ) ) {
            return $dokan_closing_time;
        }

        if ( isset( $dokan_closing_time[ $index ] ) ) {
            return $dokan_closing_time[ $index ];
        }

        return $dokan_closing_time[0]; // return the 1st index or the whole array according to your need
```

**Review by tanmayjay — CHANGES_REQUESTED** [review]

**Review by tanmayjay — APPROVED** [review]

**Review by tanmayjay — APPROVED** [review]

**Comment by Rat01047** [comment]
Issue:
Scenario:
The Opening and Closing time headers are given as Opening times* and Closing times* which does not apply for single Opening and Closing time.

Steps to Reproduce:
1. Vendor -> Dashboard -> Settings -> Store
2. Enable "Store Opening Closing Time"
3. Select day > Time selectors headers 
4. For single time selections it shows > Opening times - Closing times

Expected Result:
The Opening times* and Closing times* should be given as Opening time(s) and Closing time(s), for presenting that there are 1+ option for adding times.

Screenshot for reference:
image

**Pull request merged**